File: 16-Trie/01-Standard-Trie/java/Trie.java
public class Trie { int[][] ch; boolean[] end; int cnt=0;
    Trie(){ch=new int[200005][26];end=new boolean[200005];cnt=0;java.util.Arrays.fill(ch[0],-1);}
    void insert(String w){int n=0;for(char c:w.toCharArray()){int i=c-'a';if(ch[n][i]==-1){ch[n][i]=++cnt;java.util.Arrays.fill(ch[cnt],-1);}n=ch[n][i];}end[n]=true;}
    boolean search(String w){int n=0;for(char c:w.toCharArray()){int i=c-'a';if(ch[n][i]==-1)return false;n=ch[n][i];}return end[n];}
    public static void main(String[] args){Trie t=new Trie();t.insert("apple");System.out.println(t.search("apple"));}}

File: 16-Trie/01-Standard-Trie/cpp/trie.cpp
/* Implement Trie | LeetCode 208 | Tags: trie */
#include <bits/stdc++.h>
using namespace std;
struct TrieNode{TrieNode*ch[26]={};bool end=false;};
class Trie{TrieNode*root;public:Trie(){root=new TrieNode();}
void insert(string w){auto n=root;for(char c:w){if(!n->ch[c-'a'])n->ch[c-'a']=new TrieNode();n=n->ch[c-'a'];}n->end=true;}
bool search(string w){auto n=root;for(char c:w){if(!n->ch[c-'a'])return false;n=n->ch[c-'a'];}return n->end;}
bool startsWith(string p){auto n=root;for(char c:p){if(!n->ch[c-'a'])return false;n=n->ch[c-'a'];}return true;}};
int main(){Trie t;t.insert("apple");cout<<boolalpha<<t.search("apple")<<endl<<t.startsWith("app")<<endl<<t.search("app")<<endl;}
